clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -431,6 +444,96 @@ private static int parseAcks(String acksString) {
+     *      messages issued by the producer.
+     *
+     * @throws IllegalStateException if the TransactionalId for the producer is not set
+     *         in the configuration.
+     */
+    public void initTransactions() {
+        if (transactionState == null || !transactionState.isTransactional()) {
+            throw new IllegalStateException("Cannot call initTransactions without setting a transactional id.");
+        }
+        FutureTransactionalResult result = transactionState.initializeTransactions();
+        result.get();
+    }
+
+    /**
+     * Should be called before the start of each new transaction.
+     *
+     * @throws ProducerFencedException if another producer is with the same
+     *         transactional.id is active.
+     */
+    public void beginTransaction() throws ProducerFencedException {
+        // Set the transactional bit in the producer.
+        if (transactionState == null || !transactionState.isTransactional() || transactionState.isInTransaction()
+                || transactionState.isCompletingTransaction() || !transactionState.hasPid())
+            throw new IllegalStateException("Cannot begin a new transaction. Either transactions are not configured, " +
+                    "or there is already a transaction ongoing, or the producer is fenced.");
+        transactionState.beginTransaction();
+    }
+
+    /**
+     * Sends a list of consumed offsets to the consumer group coordinator, and also marks
+     * those offsets as part of the current transaction. These offsets will be considered
+     * consumed only if the transaction is committed successfully.
+     *
+     * This method should be used when you need to batch consumed and produced messages
+     * together, typically in a consume-transform-produce pattern.
+     *
+     * @throws ProducerFencedException if another producer with the same
+     *         transactional.id is active.
+     */
+    public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
+                                         String consumerGroupId) throws ProducerFencedException {
+        if (transactionState == null || !transactionState.isInTransaction() || transactionState.isCompletingTransaction()
+                || !transactionState.hasPid())
+            throw new IllegalStateException("Cannot send offsets to transaction. Either transactions are not enabled, or a transaction" +
+                    " isn't active right now, or a transaction is in the process of being completed, or the producer" +
+                    " has been fenced.");
+        FutureTransactionalResult result = transactionState.sendOffsetsToTransaction(offsets, consumerGroupId);
+        result.get();
+    }
+
+    /**
+     * Commits the ongoing transaction.
+     *
+     * @throws ProducerFencedException if another producer with the same
+     *         transactional.id is active.
+     */
```

Comment:
Maybe we should also document the illegal states? Same for the other public methods.